File: plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/json/Event.java
/*******************************************************************************
 * Copyright (c) 2012-2017 Codenvy, S.A.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Codenvy, S.A. - initial API and implementation
 *******************************************************************************/
package org.eclipse.che.plugin.docker.client.json;

import com.google.gson.annotations.SerializedName;

/**
 * Docker event.
 *
 * @author Alexander Garagatyi
 * @author Mykola Morhun
 */
public class Event {
    @SerializedName("status")
    private String status;
    @SerializedName("id")
    private String id;
    @SerializedName("from")
    private String from;
    @SerializedName("Type")
    private String type;
    @SerializedName("Action")
    private String action;
    @SerializedName("Actor")
    private Actor  actor;
    @SerializedName("time")
    private long   time;
    @SerializedName("timeNano")
    private long   timeNano;

    public long getTime() {
        return time;
    }

    public String getFrom() {
        return from;
    }

    public String getId() {
        return id;
    }

    public String getStatus() {
        return status;
    }

    public String getType() {
        return type;
    }

    public String getAction() {
        return action;
    }

    public Actor getActor() {
        return actor;
    }

    public long getTimeNano() {
        return timeNano;
    }

    public Event withFrom(String from) {
        this.from = from;
        return this;
    }

    public Event withId(String id) {
        this.id = id;
        return this;
    }

    public Event withStatus(String status) {
        this.status = status;
        return this;
    }

    public Event withType(String type) {
        this.type = type;
        return this;
    }

    public Event withAction(String action) {
        this.action = action;
        return this;
    }

    public Event withActor(Actor actor) {
        this.actor = actor;
        return this;
    }

    public Event withTime(long time) {
        this.time = time;
        return this;
    }

    public Event withTimeNano(long timeNano) {
        this.timeNano = timeNano;
        return this;
    }

    @Override
    public String toString() {
        return "Event{" +
               "status='" + status + '\'' +
               ", id='" + id + '\'' +
               ", from='" + from + '\'' +
               ", type='" + type + '\'' +
               ", action='" + action + '\'' +
               ", actor='" + actor + '\'' +
               ", time=" + time +
               ", timeNano=" + timeNano +
               '}';
    }
}

File: agents/go-agents/core/event/event.go
//
// Copyright (c) 2012-2017 Codenvy, S.A.
// All rights reserved. This program and the accompanying materials
// are made available under the terms of the Eclipse Public License v1.0
// which accompanies this distribution, and is available at
// http://www.eclipse.org/legal/epl-v10.html
//
// Contributors:
//   Codenvy, S.A. - initial API and implementation
//

// Package event provides lightweight primitives for event-bus-consumer communication.
package event

import "sync"

// E is an interface for event.
// Only instances of this interface can be published in Bus.
type E interface {

	// Type returns a type of this event.
	// Bus uses returned type to retrieve and notify consumers.
	Type() string
}

// Consumer is an interface for event handlers.
type Consumer interface {

	// Accept handles given event.
	Accept(event E)
}

// TmpConsumer allows consumer to be removed from bus each time after accept.
// Accept is called at least once for such consumers.
type TmpConsumer interface {
	Consumer

	// IsDone called by bus after accept, if func returns true
	// this consumer will be removed from bus.
	IsDone() bool
}

// Bus is a event-consumer communication layer.
// It manages events publications and handling, its synchronous
// and safe to use from multiple goroutines.
// Use NewBus() func to get an instance of it.
type Bus struct {
	sync.RWMutex
	consumers map[string][]Consumer
}

// ConsumerF is a function which is also consumer,
// allows to use anonymous functions as consumers.
type ConsumerF func(event E)

// Accept calls consumer function passing event to it.
func (f ConsumerF) Accept(event E) { f(event) }

// NewBus creates a new bus.
func NewBus() *Bus {
	return &Bus{consumers: make(map[string][]Consumer)}
}

// Pub publishes an event to the interested consumers.
func (bus *Bus) Pub(e E) {
	bus.RLock()
	defer bus.RUnlock()
	cons, ok := bus.consumers[e.Type()]
	if ok {
		for idx, v := range cons {
			v.Accept(e)
			if tmpCons, ok := v.(TmpConsumer); ok && tmpCons.IsDone() {
				bus.rm(e.Type(), idx)
			}
		}
	}
}

// Sub subscribes to the events of given type so
// given consumer will receive them.
func (bus *Bus) Sub(consumer Consumer, eType string) {
	bus.Lock()
	defer bus.Unlock()
	bus.consumers[eType] = append(bus.consumers[eType], consumer)
}

// SubAny does the same as Sub func, but for multiple event types.
func (bus *Bus) SubAny(consumer Consumer, types ...string) {
	bus.Lock()
	defer bus.Unlock()
	for _, t := range types {
		bus.consumers[t] = append(bus.consumers[t], consumer)
	}
}

// Rm removes given consumer from the bus, == operator is used to determine
// Returns true if consumer is removed.
func (bus *Bus) Rm(consumer Consumer) bool {
	return bus.RmIf(func(c Consumer) bool { return c == consumer })
}

// RmIf removes consumers for which predicate func returns true.
// Returns true if any consumer was removed.
func (bus *Bus) RmIf(predicate func(c Consumer) bool) bool {
	bus.Lock()
	defer bus.Unlock()
	ok := false
	for _type, candidates := range bus.consumers {
		for idx, candidate := range candidates {
			if predicate(candidate) {
				bus.rm(_type, idx)
				ok = true
			}
		}
	}
	return ok
}

// Clear assigns new map value to bus consumers and returns old ones.
// The modification of returned map won't affect bus consumers.
func (bus *Bus) Clear() map[string][]Consumer {
	bus.Lock()
	defer bus.Unlock()
	tmp := bus.consumers
	bus.consumers = make(map[string][]Consumer)
	return tmp
}

func (bus *Bus) rm(key string, idx int) {
	arr := bus.consumers[key]
	if len(arr) == 1 {
		delete(bus.consumers, key)
	} else {
		newArr := make([]Consumer, 0, len(arr)-1)
		newArr = append(newArr, arr[:idx]...)
		newArr = append(newArr, arr[idx+1:]...)
		bus.consumers[key] = newArr
	}
}
